encoding/int/encoding.go

```diff
@@ -0,0 +1,42 @@
+package int
+
+import "io"
+
+// Encoder encodes int values.
+type Encoder interface {
+	// Encode a block if ints.
+	Encode(values []int) error
+	// Bytes reutrns the encoded bytes of the encoder.
+	// The format of the byte slice is as follows:
+	//   - version:                 1 byte
+	//   - metadata length bytes:   VInt
+	//   - metadata bytes:          Variable bytes
+	//   - data bytes:              Variable bytes
+	Bytes() []byte
+	// Reset internal memory buffers manually.
+	// Otherwise, buffers are reset upon the next call to `Encode`.
+	Reset()
+}
+
+// Decoder decodes int values.
+type Decoder interface {
+	// Decode the source bytes.
+	Decode(src io.Reader) (Iterator, error)
+	// Reset internal memory buffers manually.
+	// Otherwise, buffers are reset upon the next call to `Decode`.
+	Reset()
+}
+
+// Iterator lazily produces int values decoded from a byte stream.
+type Iterator interface {
+	io.Closer
+
+	// Next returns true if there is another value
+	// in the data stream. If it returns false, err should be
+	// checked for errors.
+	Next() bool
+	// Current returns the current value in the iteration.
+	Current() int
```

Comment:
Do you need to return metadata in here?

Reply:
I was thinking that metadata should be private to the encoder.

Reply:
Yea, each encoder will prob have diff metadata implementations. I'm thinking it would be fine to leave it opaquely in the encoded byte slice since it _probably_ won't be used outside of the encoder. 

On the other hand... the encoder should return the "Block metadata" which contains meta info about that data block to be used during queries?

Reply:
Hm that feels to me should not be part of encoder's responsibilities, but rather the decoder or a component that's higher level above the decoder to know about the meta information.

What specifically is included in the meta information you are thinking about?

Reply:
Some of the meta information will *only* be used by the decoder to decode the data like you said. That information will not be exposed and will live in the head of the `[]byte` slice returned by the encoder `Bytes()` method.

Some other info will be stored as block metadata for querying purposes. For example:
- cardinality
- min/max
- doc id start/stop
- unix ts start/stop

Also updated my segment file design doc: https://docs.google.com/document/d/1gMsn1VCiZELVm_tDSOy4a9pbvn5e1gXO7-LyN2Ma2Oo/edit

Reply:
Interesting, what's the distinction between metadata stored as part of the encoded bytes and the metadata stored as block metadata exposed for querying? How do you plan to store the block metadata?

Reply:
Discussed offline. Encoder is responsible for parsing out metadata info from the actual data.